src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/RazorHelperUpdater.cs

```diff
@@ -0,0 +1,153 @@
+                if (await _helperMatcher.HasHelperAsync(input).ConfigureAwait(false))
+                {
+                    // If a @helper is present, log it to the console and store the path to return later
+                    var path = input.Source.FilePath;
+                    inputsWithHelpers.Add(path);
+                }
+            }
+
+            // Log how many Razor documents had @helpers and return a corresponding FileUpdaterResult
+            _logger.LogInformation("Found @helper functions in {Count} documents", inputsWithHelpers.Count);
+            return new FileUpdaterResult(inputsWithHelpers.Any(), inputsWithHelpers);
+        }
+
+        /// <summary>
+        /// Upgrades a set of RazorCodeDocuments to replace @helper functions with local methods.
+        /// </summary>
+        /// <param name="context">The upgrade context being upgraded.</param>
+        /// <param name="inputs">The Razor code documents to upgrade.</param>
+        /// <param name="token">A cancellation token.</param>
+        /// <returns>A FileUpdaterResult indicating whether the opertaion succeeded and which documents (if any) were updated.</returns>
+        public async Task<IUpdaterResult> ApplyAsync(IUpgradeContext context, ImmutableArray<RazorCodeDocument> inputs, CancellationToken token)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            // Track which documents are updated for inclusion in the updater result returned later
+            var updatedDocuments = new List<string>();
+            foreach (var document in inputs)
+            {
+                var path = document.Source.FilePath;
+                var helperReplacements = await _helperMatcher.GetHelperReplacementsAsync(document).ConfigureAwait(false);
+
+                if (helperReplacements.Any())
+                {
+                    // If the document contains a @helper, construct the new document text by removing the replacement's old text
+                    // and inserting the new text in its place.
+                    updatedDocuments.Add(path);
+                    var newText = new StringBuilder(File.ReadAllText(path));
+                    _logger.LogInformation("Replacing {HelperCount} @helper functions in {FilePath}", helperReplacements.Count(), path);
+                    foreach (var helper in helperReplacements.OrderByDescending(h => h.Offset))
+                    {
+                        helper.Apply(newText);
+                    }
+
+                    // Check whether the document has a using statement for HelperResult's namespace (Microsoft.AspNetCore.Mvc.Razor) and
+                    // add one if none is present. Again, this is done with string manipulation since there's no good API for updating
+                    // Razor syntax.
+                    if (!HasMvcRazorUsing(document))
```

Comment:
Could we check if there's an `_Imports.cshtml` that would have this?

Reply:
That will happen automatically. That's why `HasMvcRazorUsing` is looking at the generated C# - that source will include source from imported documents in addition to the source from the primary document.